File: internal/webhook/format.go
package webhook

import (
	"fmt"

	"github.com/dylsteck/coolify-doctor/internal/coolify"
	"github.com/dylsteck/coolify-doctor/internal/telegram"
)

// Format renders a Coolify webhook payload into a Telegram HTML message. Raw
// body is passed through for the unknown-event fallback so nothing is silently
// dropped.
func Format(e coolify.Event, raw []byte) string {
	icon := "✅"
	if !e.Success {
		icon = "❌"
	}

	switch e.Event {
	case "deployment_success", "deployment_failed":
		title := "Deployment succeeded"
		if e.Event == "deployment_failed" {
			title = "Deployment failed"
		}
		return telegram.JoinLines(
			fmt.Sprintf("%s <b>%s</b>", icon, telegram.Esc(title)),
			appLine(e),
			telegram.Link("View deployment", e.DeploymentURL),
		)

	case "status_changed":
		return telegram.JoinLines(
			fmt.Sprintf("⚠️ <b>%s</b>", telegram.Esc(orDefault(e.Message, "Status changed"))),
			appLine(e),
			telegram.Link("Open", e.URL),
		)

	case "backup_success", "backup_failed":
		title := "Backup succeeded"
		if e.Event == "backup_failed" {
			title = "Backup failed"
		}
		lines := []string{
			fmt.Sprintf("%s <b>%s</b>", icon, telegram.Esc(title)),
			dbLine(e),
		}
		if e.ErrorOutput != "" {
			lines = append(lines, fmt.Sprintf("<pre>%s</pre>", telegram.Esc(telegram.Truncate(e.ErrorOutput, 500))))
		}
		lines = append(lines, telegram.Link("Open", e.URL))
		return telegram.JoinLines(lines...)

	case "task_success", "task_failed":
		title := "Scheduled task succeeded"
		if e.Event == "task_failed" {
			title = "Scheduled task failed"
		}
		lines := []string{fmt.Sprintf("%s <b>%s</b>", icon, telegram.Esc(title))}
		if e.TaskName != "" {
			lines = append(lines, fmt.Sprintf("Task: <b>%s</b>", telegram.Esc(e.TaskName)))
		}
		if e.Output != "" {
			lines = append(lines, fmt.Sprintf("<pre>%s</pre>", telegram.Esc(telegram.Truncate(e.Output, 500))))
		}
		lines = append(lines, telegram.Link("Open", e.URL))
		return telegram.JoinLines(lines...)

	case "server_reachable", "server_unreachable", "server_disk_usage", "server_patch":
		return telegram.JoinLines(
			fmt.Sprintf("%s <b>%s</b>", icon, telegram.Esc(telegram.PrettyEvent(e.Event))),
			serverLine(e),
			telegram.Esc(e.Message),
		)

	case "traefik_version_outdated":
		return telegram.JoinLines(
			"⚠️ <b>Traefik outdated</b>",
			telegram.Esc(e.Message),
		)

	case "docker_cleanup_success", "docker_cleanup_failed":
		return telegram.JoinLines(
			fmt.Sprintf("%s <b>%s</b>", icon, telegram.Esc(telegram.PrettyEvent(e.Event))),
			serverLine(e),
			telegram.Esc(e.Message),
		)

	case "test":
		return fmt.Sprintf("🧪 <b>Test webhook</b>\n%s", telegram.Esc(e.Message))
	}

	return fmt.Sprintf(
		"📨 <b>%s</b>\n%s\n<pre>%s</pre>",
		telegram.Esc(orDefault(e.Event, "unknown_event")),
		telegram.Esc(e.Message),
		telegram.Esc(telegram.Truncate(string(raw), 800)),
	)
}

func appLine(e coolify.Event) string {
	parts := []string{fmt.Sprintf("<b>%s</b>", telegram.Esc(orDefault(e.ApplicationName, "app")))}
	if e.Project != "" {
		parts = append(parts, telegram.Esc(e.Project))
	}
	if e.Environment != "" {
		parts = append(parts, telegram.Esc(e.Environment))
	}
	return telegram.JoinInline(parts...)
}

func dbLine(e coolify.Event) string {
	parts := []string{fmt.Sprintf("<b>%s</b>", telegram.Esc(orDefault(e.DatabaseName, "database")))}
	if e.DatabaseType != "" {
		parts = append(parts, telegram.Esc(e.DatabaseType))
	}
	if e.Frequency != "" {
		parts = append(parts, "every "+telegram.Esc(e.Frequency))
	}
	return telegram.JoinInline(parts...)
}

func serverLine(e coolify.Event) string {
	if e.ServerName == "" {
		return ""
	}
	return fmt.Sprintf("Server: <b>%s</b>", telegram.Esc(e.ServerName))
}

func orDefault(s, d string) string {
	if s == "" {
		return d
	}
	return s
}

File: src/coolify/format.ts
import type { CoolifyEvent } from "./event.js";
import { esc, joinInline, joinLines, link, prettyEvent, truncate } from "../telegram/html.js";

function orDefault(s: string, d: string): string {
  return s === "" ? d : s;
}

function appLine(e: CoolifyEvent): string {
  const parts = [`<b>${esc(orDefault(e.application_name ?? "", "app"))}</b>`];
  if (e.project) parts.push(esc(e.project));
  if (e.environment) parts.push(esc(e.environment));
  return joinInline(...parts);
}

function dbLine(e: CoolifyEvent): string {
  const parts = [`<b>${esc(orDefault(e.database_name ?? "", "database"))}</b>`];
  if (e.database_type) parts.push(esc(e.database_type));
  if (e.frequency) parts.push("every " + esc(e.frequency));
  return joinInline(...parts);
}

function serverLine(e: CoolifyEvent): string {
  if (!e.server_name) return "";
  return `Server: <b>${esc(e.server_name)}</b>`;
}

/** Renders Coolify webhook JSON into Telegram HTML. */
export function formatEvent(e: CoolifyEvent, raw: string): string {
  const icon = e.success ? "✅" : "❌";

  switch (e.event) {
    case "deployment_success":
    case "deployment_failed": {
      const title = e.event === "deployment_failed" ? "Deployment failed" : "Deployment succeeded";
      return joinLines(
        `${icon} <b>${esc(title)}</b>`,
        appLine(e),
        link("View deployment", e.deployment_url ?? ""),
      );
    }
    case "status_changed":
      return joinLines(
        `⚠️ <b>${esc(orDefault(e.message, "Status changed"))}</b>`,
        appLine(e),
        link("Open", e.url ?? ""),
      );
    case "backup_success":
    case "backup_failed": {
      const title = e.event === "backup_failed" ? "Backup failed" : "Backup succeeded";
      const lines = [`${icon} <b>${esc(title)}</b>`, dbLine(e)];
      if (e.error_output) {
        lines.push(`<pre>${esc(truncate(e.error_output, 500))}</pre>`);
      }
      lines.push(link("Open", e.url ?? ""));
      return joinLines(...lines);
    }
    case "task_success":
    case "task_failed": {
      const title = e.event === "task_failed" ? "Scheduled task failed" : "Scheduled task succeeded";
      const lines = [`${icon} <b>${esc(title)}</b>`];
      if (e.task_name) lines.push(`Task: <b>${esc(e.task_name)}</b>`);
      if (e.output) lines.push(`<pre>${esc(truncate(e.output, 500))}</pre>`);
      lines.push(link("Open", e.url ?? ""));
      return joinLines(...lines);
    }
    case "server_reachable":
    case "server_unreachable":
    case "server_disk_usage":
    case "server_patch":
      return joinLines(
        `${icon} <b>${esc(prettyEvent(e.event))}</b>`,
        serverLine(e),
        esc(e.message),
      );
    case "traefik_version_outdated":
      return joinLines("⚠️ <b>Traefik outdated</b>", esc(e.message));
    case "docker_cleanup_success":
    case "docker_cleanup_failed":
      return joinLines(
        `${icon} <b>${esc(prettyEvent(e.event))}</b>`,
        serverLine(e),
        esc(e.message),
      );
    case "test":
      return `🧪 <b>Test webhook</b>\n${esc(e.message)}`;
    default:
      return joinLines(
        `📨 <b>${esc(orDefault(e.event, "unknown_event"))}</b>`,
        esc(e.message),
        `<pre>${esc(truncate(raw, 800))}</pre>`,
      );
  }
}
